be/src/serde/protobuf_serde.h

```diff
@@ -14,29 +14,71 @@ class RowDescriptor;
 }
 
 namespace starrocks::serde {
+constexpr double EncodeRatioLimit = 0.9;
+constexpr uint32_t EncodeSamplingNum = 5;
+
+// EncodeContext adaptively adjusts encode_level according to the compression ratio. In detail,
+// for every _frequency chunks, if the compression ratio for the first EncodeSamplingNum chunks is less than
+// EncodeRatioLimit, then encode the rest chunks, otherwise not.
+
+class EncodeContext {
+public:
+    static std::shared_ptr<EncodeContext> get_encode_context_shared_ptr(const int col_num, const int encode_level) {
+        return encode_level == 0 ? nullptr : std::make_shared<EncodeContext>(col_num, encode_level);
+    }
+    EncodeContext(const int col_num, const int encode_level);
+    // update encode_level for each column
+    void update(const int col_id, uint64_t mem_bytes, uint64_t encode_byte);
+
+    int get_encode_level(const int col_id) {
+        DCHECK(_session_encode_level != 0);
+        return _column_encode_level[col_id];
+    }
+
+    int get_session_encode_level() {
+        DCHECK(_session_encode_level != 0);
+        return _session_encode_level;
+    }
+
+    void set_encode_levels_in_pb(ChunkPB* const res);
+
+private:
+    // if encode ratio < EncodeRatioLimit, encode it, otherwise not.
+    void _adjust(const int col_id);
+    const int _session_encode_level;
+    uint64_t _times = 0;
+    uint64_t _frequency = 64;
```

Comment:
what's the meaning of `_frequency` in here?

Reply:
for every _frequency chunks, we will check the encoding ratio to decide encoding or not.